spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/BootstrapApplicationListener.java

```diff
@@ -103,7 +102,11 @@ public void onApplicationEvent(ApplicationEnvironmentPreparedEvent event) {
 		if (context == null) {
 			context = bootstrapServiceContext(environment, event.getSpringApplication(),
 					configName);
+			ConfigurableApplicationContext bootstrapContext = context;
+			event.getSpringApplication().addListeners((ApplicationListener<ApplicationFailedEvent>) afe ->
```

Comment:
This is very slow at runtime. It's better to have a `SmartApplicationListener` so the listener class doesn't have to be introspected.

Reply:
ok, but is the approach the right one?

Reply:
Ok, I've updated to use a `SmartApplicationListener`